internal/services/elastic/elastic.go

```diff
@@ -0,0 +1,232 @@
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"slices"
+	"strconv"
+	"time"
+
+	"github.com/DIMO-Network/device-data-api/internal/config"
+	"github.com/DIMO-Network/device-data-api/internal/constants"
+	"github.com/elastic/go-elasticsearch/v8"
+	"github.com/elastic/go-elasticsearch/v8/typedapi/core/search"
+	"github.com/elastic/go-elasticsearch/v8/typedapi/types"
+	"github.com/rs/zerolog"
+	"github.com/tidwall/gjson"
+)
+
+const (
+	docByInterval  = "documents_by_interval"
+	singleDoc      = "select_single_doc"
+	timeField      = "time"
+	subjectField   = "subject"
+	defaultRetries = 5
+)
+
+// Service is a service for performing queries on elastic.
+type Service struct {
+	settings *config.Settings
+	log      *zerolog.Logger
+	esClient *elasticsearch.TypedClient
+}
+
+// New creates a newly configured elastic service.
+func New(settings *config.Settings, logger *zerolog.Logger, caCert []byte) (*Service, error) {
+	esConfig := elasticsearch.Config{
+		Addresses:  []string{settings.ElasticSearchAnalyticsHost},
+		Username:   settings.ElasticSearchAnalyticsUsername,
+		Password:   settings.ElasticSearchAnalyticsPassword,
+		CACert:     caCert,
+		MaxRetries: defaultRetries,
+	}
+	es8Client, err := elasticsearch.NewTypedClient(esConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create elastic client: %w", err)
+	}
+	eClient := Service{
+		esClient: es8Client,
+		log:      logger,
+		settings: settings,
+	}
+	return &eClient, nil
+}
+
+// EsClient returns the underlying elastic client used by the service.
+//
+// This function exists to maintain previous behavior for endpoint who are not yet using the elastic service.
+func (s *Service) EsClient() *elasticsearch.TypedClient {
+	return s.esClient
+}
+
+// GetHistoryParams defines the parameters for retrieving history from elastic.
+type GetHistoryParams struct {
+	// StartTime is the start time of the history.
+	StartTime time.Time
+	// EndTime is the end time of the history.
+	EndTime time.Time
+	// DeviceID is the ID of the device.
+	DeviceID string
+	// PrivilegeIDs is the list of privilege IDs that the user has.
+	PrivilegeIDs []int64
+	// Buckets is the number of time intervals to divide the history into.
+	Buckets int
+}
+
+// GetHistory retrieves the history of a device from elastic. The history is divided into buckets and one data point is selected from each bucket.
+// The data points are selected by the first data point in the bucket that has a vehicle field.
+// The result is a list of data points.
+func (s *Service) GetHistory(ctx context.Context, params GetHistoryParams) ([]any, error) {
+	// set default values for the params
+	setDefaultHistoryParams(&params)
+
+	// build the query
+	req := buildHistoryQuery(params)
+
+	// log requests
+	reqb, err := json.Marshal(req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal request: %w", err)
+	}
+	s.log.Info().RawJSON("request", reqb).Str("index", s.settings.DeviceDataIndexNameV2).Msg("V2 request")
```

Comment:
This could get very spammy, but it's good for now.

Reply:
I left it here because it was part of the original implementation, but I do not love that we have to marshal the query just to log it. If you are okay with it, I would prefer to remove this log.

Reply:
Raise it to Debug level maybe?